### ISSUE TITLE
Kestrel server crashes with System.BadImageFormatException: Bad IL format.

### ISSUE THREAD

**Issue by streepteaze** [aspnet/KestrelHttpServer#2680]
ASP.NET Core 2.1.0 web application deployed to Windows Azure App Service randomly crashes in production environment and server stops handling resuests, just sending 502.3 errors:
Response: The specified CGI application encountered an error and the server terminated the process.

The problem occurs multiple times in a day, the only way to recover it is to restart Azure App Service. CPU and Memory usage are ok before the crash, no any additional errors or failed requests have been found before crashes.

Application is using IdentityServer4 and compiled in Release (Any CPU) default configuration.

From event logs I can see an exception causing process terminating:

```xml
<Event>
  <System>
    <Provider Name=".NET Runtime"/>
    <EventID>1026</EventID>
    <Level>1</Level>
    <Task>0</Task>
    <Keywords>Keywords</Keywords>
    <TimeCreated SystemTime="2018-06-20T19:15:11Z"/>
    <EventRecordID>1163393500</EventRecordID>
    <Channel>Application</Channel>
    <Computer>RD0003FF55B38D</Computer>
    <Security/>
  </System>
  <EventData>
    <Data>
Application: dotnet.exe CoreCLR Version: 4.6.26515.7 Description: The process was terminated due to an unhandled exception. Exception Info: System.BadImageFormatException: Bad IL format.

at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection.ProcessReceives()
at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.<>c.<.cctor>b__9_0(Object s)
at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---

at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketAwaitable.<>c.<Complete>b__12_0(Object state)
at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.IOQueue.DoWork()
at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.IOQueue.<>c.<.cctor>b__8_0(Object s) at System.Threading.QueueUserWorkItemCallback.<>c.<.cctor>b__6_0(Object state)
at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---

at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
at System.Threading.QueueUserWorkItemCallback.ExecuteWorkItem()
at System.Threading.ThreadPoolWorkQueue.Dispatch()
at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
    </Data>
  </EventData>
</Event>
```

Azure App Service dotnet info:
```
> dotnet --info
D:\home\site\wwwroot
.NET Core SDK (reflecting any global.json):
 Version:   2.1.300
 Commit:    32f29b6eb9

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.14393
 OS Platform: Windows
 RID:         win10-x86
 Base Path:   D:\Program Files (x86)\dotnet\sdk\2.1.300\

Host (useful for support):
  Version: 2.1.0
  Commit:  caa7b7e2ba

.NET Core SDKs installed:
  1.1.8 [D:\Program Files (x86)\dotnet\sdk]
  2.1.101 [D:\Program Files (x86)\dotnet\sdk]
  2.1.300 [D:\Program Files (x86)\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.0-rc1-final [D:\Program Files (x86)\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.0 [D:\Program Files (x86)\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.0-rc1-final [D:\Program Files (x86)\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.0 [D:\Program Files (x86)\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 1.0.10 [D:\Program Files (x86)\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.7 [D:\Program Files (x86)\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.6 [D:\Program Files (x86)\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.0-rc1 [D:\Program Files (x86)\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.0 [D:\Program Files (x86)\dotnet\shared\Microsoft.NETCore.App]
```

**Comment by Tratcher** [comment]
This looks like https://github.com/dotnet/coreclr/issues/18323. Give the workaround a try and let us know.

**Comment by streepteaze** [comment]
Hey, @tratcher thank you for pointing out the issue. The nature looks similar, however the stack trace which leads to process crash for me looks always the same and not related to EF.

I just disabled AI Profiler Web Job and AI Profiler Web Job monitoring, so will give you an update if anything will change in next few days.

I also has found a couple of other System.BadImageFormatException exceptions in logs, however they don't lead to process crash, just to 'Connection processing ended abnormally' errors.

Here is some of them: https://gist.github.com/streepteaze/e150c8d6eb431f2e90664c671b3079b2

**Comment by streepteaze** [comment]
Unfortunately this workaround with disabling profiler didn't help, so I'm still experiencing 2 - 7 process crashes a day and ~100 non-fatal `System.BadImageFormatException` exceptions daily.

**Comment by benaadams** [comment]
Does it look like https://github.com/dotnet/coreclr/issues/18448 ?

**Comment by streepteaze** [comment]
Crashes appear only when the project is deployed to Azure App Service, and it has no references to New Relic or App Dynamics profilers.

**Comment by streepteaze** [comment]
After redeploying an app, I was able to catch logs right before and after process crash, so possibly it could be relevant and will help to identify the problem: https://gist.github.com/streepteaze/abea8fae92c045c128601c3603fa9d86

**Comment by muratg** [comment]
@streepteaze What exactly did you disable?

**Comment by muratg** [comment]
cc @shirhatti

**Comment by streepteaze** [comment]
@muratg 

![image](https://user-images.githubusercontent.com/772082/42454617-9e7c1092-8398-11e8-9feb-facd5652ff23.png)

**Comment by Eilon** [comment]
@muratg / @pakrym / @halter73 - does Kestrel do any wacky IL ref emit at runtime that might cause magical bad IL to show up? Or could this still be caused by https://github.com/dotnet/coreclr/issues/18448 somehow? (Which looks like it's planned to be fixed in the next patch.)

**Comment by pakrym** [comment]
Only DI does, considering exception is in at `Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection.ProcessReceives()` still looks like  dotnet/coreclr#18448 to me.

**Comment by streepteaze** [comment]
Hi, all! After some experiments I can confirm that updating Application Insights site extension to latest 2.6.5 version and switching off profiler when reconnecting Application Insights resource to App Service solved the problem.

![image](https://user-images.githubusercontent.com/772082/42806690-ac553b7e-89b7-11e8-8b5c-21f0bc8c2ae7.png)

However it still confuses me, as App Insights UI reports that profiler is enabled.

![image](https://user-images.githubusercontent.com/772082/42807103-9e2d748e-89b8-11e8-8e22-af0bd342a9ca.png)

**Comment by mojohub** [comment]
Not sure if it helps but after a few production crashes in my Azure Web App I managed to reproduce the System.BadImageFormatException by multi clicking the submit button on one of my forms many times which then displayed "**The specified CGI application encountered an error and the server terminated the process.**" in the browser.

In the application event log of my web app i have the following entry...

> Application: dotnet.exe
> CoreCLR Version: 4.6.26628.5
> Description: The process was terminated due to an unhandled exception.
> Exception Info: System.BadImageFormatException: Bad IL format.
>    at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection.ProcessSends()
>    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.<>c.<.cctor>b__9_0(Object s)
>    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
>    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
>    at System.Runtime.CompilerServices.ValueTaskAwaiter.<>c.<.cctor>b__11_1(Object state)
>    at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.IOQueue.DoWork()
>    at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.IOQueue.<>c.<.cctor>b__8_0(Object s)
>    at System.Threading.QueueUserWorkItemCallback.<>c.<.cctor>b__6_0(Object state)
>    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
>    at System.Threading.QueueUserWorkItemCallback.ExecuteWorkItem()
>    at System.Threading.ThreadPoolWorkQueue.Dispatch()
>    at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()

I will be going through my app and adding javascript to disable the submit button once clicked all my forms to help prevent this happening again in future. However should the runtime be able to handle this scenario without crashing?

**Comment by halter73** [comment]
@mojohub Have you tried updating the App Insights site extension and disabling profiler like @streepteaze did?

**Comment by mojohub** [comment]
@halter73 Yes I have the latest App Insights extension installed (2.6.5) and I also disabled the profiler from running as per findings from @streepteaze above. I did not have another crash since doing that however I was able to force the crash myself by smashing the save button on one of my forms (ie clicking it a dozen times or so in quick succession).

I came across this because I found 10 duplicate records in my database that were created by one user all within a few seconds of each other and this was just before the last production crash. The only way this could happen was if the user was trigger happy so when I tested this myself in my dev environment it sure enough crashed the system. I cannot say whether disabling profiler made any difference or i'm just lucky that nobody has been so trigger happy in production lately.

If you want me to generate any other info or dumps from doing this just let me know?
